parachain/pallets/basic-channel/src/inbound/benchmarking.rs

```diff
@@ -0,0 +1,255 @@
+//! BasicInboundChannel pallet benchmarking
+
+#![cfg(feature = "runtime-benchmarks")]
+
+use super::*;
+
+use frame_system::{RawOrigin, Pallet, self, EventRecord};
+use frame_benchmarking::{benchmarks, whitelisted_caller, impl_benchmark_test_suite};
+use hex_literal::hex;
+use sp_std::convert::TryInto;
+
+use artemis_core::{ChannelId, Message, MessageId, Proof};
+use artemis_ethereum::{Log, Header};
+
+#[allow(unused_imports)]
+use crate::inbound::Module as BasicInboundChannel;
+
+fn assert_last_event<T: Config>(system_event: <T as frame_system::Config>::Event) {
+	let events = Pallet::<T>::events();
+	// compare to the last event record
+	let EventRecord { event, .. } = &events[events.len() - 1];
+	assert_eq!(event, &system_event);
+}
+
+fn make_header_chain(parent: Header, length: u32) -> Vec<Header> {
+	let mut chain = vec![parent.clone()];
+	(0..length).fold(parent, |ancestor, _| {
+		let mut header: Header = Default::default();
+		header.parent_hash = ancestor.compute_hash();
+		header.number = ancestor.number.saturating_add(1);
+		header.difficulty = 1.into();
+		chain.push(header.clone());
+		header
+	});
+
+	chain
+}
+
+// This collection of benchmarks should include a benchmark for each
+// call dispatched by the channel, i.e. each "app" pallet function
+// that can be invoked by MessageDispatch. The most expensive call
+// should be used in the `submit` benchmark.
+//
+// We rely on configuration via chain spec of the app pallets because
+// we don't have access to their storage here.
+benchmarks! {
+	// Benchmark `submit` extrinsic under worst case conditions:
+	// * `submit` dispatches the DotApp::unlock call
+	// * `unlock` call successfully unlocks DOT
+	submit {
+		let h in 0 .. 20;
```

Comment:
The intention is to add a parameter to `Verifier` to control this upper limit in another PR

Reply:
Actually, @vgeddes do you have any thoughts on this? I'm going back and forth on the best solution. `Verifier::verify` has a very high upper limit for storage reads, i.e. 5000. It's currently limited by the pruning range (see [`FINALIZED_HEADERS_TO_KEEP`](https://github.com/Snowfork/polkadot-ethereum/blob/7bdca4f73328a2089ccbd9a0183884a31cbf8b83/parachain/pallets/verifier-lightclient/src/lib.rs#L37)). So a message from 5000 headers ago will trigger an iteration over 5000 ancestors to ensure it's in the finalized chain.

If I were to set `submit`'s weight to the current upper limit, that's unreasonably high. Even if I implement a refund mechanism. So I'm considering the following options:
1. Simply reduce `FINALIZED_HEADERS_TO_KEEP` to a value like 20. This should be fine if messages are always forwarded as soon as their header is finalized. But it feels very aggressive.
2. Let pallets pass `max_header_age` to `Verify::verify`. So each pallet declares its own upper limit.
3. Add `FinalizedHeaderExpiresAfter` config parameter to lightclient pallet. It's worse than (1) in my opinion since we're storing a bunch of headers that we don't use

Reply:
eth is 6484.00 per day i think, so 5000 is like max 20 hours downtime

Reply:
why does Verify::verify iterate over storage reads to find the header? could it just accept the exact block number and just get that one out?

Reply:
The 5000 means we need messages to be forwarded to us within 5000 blocks of when a message's block was finalized. The implication for downtime is we can't have more than 20 hours asymmetrical downtime, i.e. we receive headers but we don't receive their messages. I don't think we can recover from that. If we don't receive both headers and messages, that's fine. Nothing will be pruned and we'll continue where we left off when forwarding is restored.

W.r.t. why we iterate: An old block's existence in storage doesn't indicate it's in the canonical chain because we don't prune stale siblings on finalization. Pruning happens only at the tail of 5000 headers. This comes from parity-bridges-common (see https://github.com/paritytech/parity-bridges-common/blob/master/modules/ethereum/src/lib.rs#L957-L963). Instead of iterating, we could set a flag in storage when a header is finalized. At a glance I think that will work. But the fact that parity-bridges-common didn't implement something like that is conspicuous to me... Shall we give it a shot anyway?

Reply:
yeh, the flag sounds like a good idea to me - i guess it means slightly higher initial costs and an extra field being stored when finalizing but then o(1) message verification for older messages and could do much later prunes, like 50k blocks?

i suppose if the expectation is that message relay will happen as soon as a block is finalized 99% of the time, then the flag could be more expensive overall longer term due to the overhead? maybe that's why they didn't add it?

@tomusdrw do you have any insight into why this was or wasn't done in parity-bridges-common?

Reply:
The linked `ethereum` crate is dealing with PoA chain, which does have a notion of finality, but it's not necessarily predictable. I.e. the fact that there has been many blocks on top of some block doesn't necessarily indicate finality. We don't use a flag there to prevent a situation where we have to mark as finalized a very long fork.
With PoW I think it's definitely a good idea to have a flag here, since the notion of "finality" is quite predictable - i.e. you just wait for a bunch of confirmations, so with a header import you always finalized at most one header.
I don't think the extra storage cost for the flag is worth to be considered TBH.

Just a random thought that I had now: perhaps it's worth to implement the storage as a ring buffer?

Reply:
Thanks @tomusdrw. That's helpful context. I'm going to add the flag.

W.r.t. a ring buffer - I think that would simplify the pallet a lot. The behaviour of the ring buffer would be unexpected though, but it probably doesn't matter. By "unexpected" I mean it cannot guarantee storage of exactly `size` block heights - no more, no less. At least, that would be my expectation of a ring buffer in this context.

Reply:
> By "unexpected" I mean it cannot guarantee storage of exactly size block heights - no more, no less. 

Ah, I was just writing a comment to ask you to elaborate on why it's not possible, but I think I know what you mean - let me get a confirmation though. So while we know that finality is achieved in `confirmations` number of blocks, we can also specify the `proof_window` (number of finalized blocks to keep) and then set the ring buffer `size = confirmations + proof_window`. However there can be many forks in the network, so we potentially need to store multiple headers under each ring buffer slot (i.e. `Vec<(Header, IsFinalized)>`). I think it's still interesting to see how big that `Vec` would be historically in the worst case, my gut says it's like 2-3 entries at most, which is not huge for decode/encode and a lookup. We could also use an enum:
```rust
enum Headers {
  Finalized(Header),
  Unfinalized(Vec<Header>),
}
```
So instead of marking the header with a flag we simply switch from vec-like enum variant to single-value enum varing in the ring buffer slot.

Reply:
> I think it's still interesting to see how big that Vec would be historically in the worst case, my gut says it's like 2-3 entries at most, which is not huge for decode/encode and a lookup

Sounds about right. I'm guessing that the ring buffer will save us 1 or 2 storage writes for the average case, and 8 in the worst case.

Just to clarify, when I said "cannot guarantee storage..." I meant that we cannot guarantee it while maintaining header insertion as an O(1) operation. I was imagining the headers are stored individually, and the ring buffer stores only pointers to those headers. So whenever you insert a header, you have to delete a variable number of stale headers using individual storage writes until you hit a deletion limit. However, your proposal of having a ring buffer slot contain the headers themselves, not pointers to the headers, means that insertion is trivially O(1) in terms of storage writes since you delete 1 slot max. It's not O(1) in terms of encoding/decoding but that's a great tradeoff for perf - at least for the `import_header` extrinsic, which is by far the most expensive.

Reply:
The ring buffer has some other annoyances, like not being able to look up a header by hash. So clients always need to know both block height + hash. I'll go think about this some more

Reply:
i don't think we should implement ring buffer now but yeh seems like it may provide some optimizations longer term, though with extra complexities